include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,255 @@
+	 * change over future versions of this API
+	 */
+	CHARGER_PROP_CUSTOM_BEGIN = CHARGER_PROP_COMMON_COUNT + 1,
+	/** Reserved to demark end of valid enum properties */
+	CHARGER_PROP_MAX = UINT16_MAX,
+};
+
+/**
+ * @brief External supply states
+ */
+enum charger_online {
+	/** External supply not presented */
+	CHARGER_ONLINE_OFFLINE = 0,
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Charger prop data types
+ */
+enum charger_type {
+	CHARGER_TYPE_ONLINE,
+	CHARGER_TYPE_BOOL,
+	CHARGER_TYPE_STATUS,
+	CHARGER_TYPE_UINT,
+};
+
+/**
+ * @brief A representation of a charger property value
+ */
+struct charger_value {
+	enum charger_type type;
+	union {
+		/* For boolean flags */
+		bool b;
+		/* For unsigned integer values */
+		unsigned int uint;
+		/* For CHARGER_PROP_ONLINE values */
+		enum charger_online online;
+		/* For CHARGER_PROP_STATUS values */
+		enum charger_status status;
+	} value;
+};
+
+#define CHARGER_VALUE_INIT_BOOL(_value)                                                            \
+	{                                                                                          \
+		.type = CHARGER_TYPE_BOOL, .value = {.b = (_value) }                               \
+	}
+#define CHARGER_VALUE_INIT_INT(_value)                                                             \
+	{                                                                                          \
+		.type = CHARGER_TYPE_UINT, .value = {.uint = (_value) }                            \
+	}
+#define CHARGER_VALUE_INIT_ONLINE(_value)                                                          \
+	{                                                                                          \
+		.type = CHARGER_TYPE_ONLINE, .value = {.online = (_value) }                        \
+	}
+#define CHARGER_VALUE_INIT_STATUS(_value)                                                          \
+	{                                                                                          \
+		.type = CHARGER_TYPE_STATUS, .value = {.status = (_value) }                        \
+	}
+
+#define CHARGER_ASSERT_TYPE(_v, _type) __ASSERT_NO_MSG((_v)->type == (_type))
```

Comment:
I'd like to hear what others think of "type safety" solely relying on `CONFIG_ASSERT=y`.

Reply:
This is not a blocker, by the way.

Reply:
I think @aaronemassey had some thoughts.

It is Friday afternoon before a holiday weekend here in the US, so I'm not sure how much a response will come from here prior to the 3.4 merge window closing.

Reply:
No worries. Changing from ASSERT-based type validation to run-time type validation can be done in another PR, if deemed necessary.

Reply:
Ahh, I understand better why that switch table was in the header now. It's so we can do these actions. This is largely why I'm more of a fan of the union member approach as it's compile-time checked.

I'd still rather move the switch statement checking the value types into the driver and do runtime checking so we don't add binary size at every call-site.

Reply:
I think we're all in agreement on the move from the API-level to the driver-level approach.